commit 984ae3529bd00eaa1b6d62e404a5c64b14ac05ed
Author: dann frazier <dannf@hp.com>
Date:   Thu Feb 11 17:07:25 2010 -0700

    [Adjusted to apply to Debian's 2.6.26 by dann frazier <dannf@debian.org>]
    
      commit 51246bfd189064079c54421507236fd2723b18f3
      Author: Thomas Gleixner <tglx@linutronix.de>
      Date:   Tue Feb 2 11:40:27 2010 +0100
    
          futex: Handle user space corruption gracefully

diff --git a/kernel/futex.c b/kernel/futex.c
index ec84da5..a316902 100644
--- a/kernel/futex.c
+++ b/kernel/futex.c
@@ -630,6 +630,13 @@ static int wake_futex_pi(u32 __user *uaddr, u32 uval, struct futex_q *this)
 	if (!pi_state)
 		return -EINVAL;
 
+	/*
+	 * If current does not own the pi_state then the futex is
+	 * inconsistent and user space fiddled with the futex value.
+	 */
+	if (pi_state->owner != current)
+		return -EINVAL;
+
 	spin_lock(&pi_state->pi_mutex.wait_lock);
 	new_owner = rt_mutex_next_owner(&pi_state->pi_mutex);
 
